app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb

```diff
@@ -1,65 +1,115 @@
 class ManageIQ::Providers::Azure::CloudManager::EventCatcher::Stream
-  #
-  # Creates an event monitor
+  SELECT_FIELDS = %w[
+    authorization
+    correlationId
+    description
+    eventDataId
+    eventName
+    eventTimestamp
+    operationName
+    resourceGroupName
+    resourceProviderName
+    resourceId
+    resourceType
+  ].join(',').freeze
+
+  # Creates an event monitor. Used internally by the Runner.
   #
   def initialize(ems)
     @ems = ems
     @collecting_events = false
     @since = nil
   end
 
-  # Start capturing events
+  # Sets a boolean used by the +each_batch+ method that indicates
+  # that events should start/keep being captured.
+  #
   def start
     @collecting_events = true
   end
 
-  # Stop capturing events
+  # Sets a boolean used by the +each_batch+ method that indicates
+  # that events should stop being captured.
+  #
   def stop
     @collecting_events = false
   end
 
+  # Used internally by the Runner#monitor_events method.
+  #
   def each_batch
     while @collecting_events
-      yield get_events.collect { |e| JSON.parse(e) }
+      yield get_events.collect(&:to_hash)
     end
   end
 
   private
 
+  # Get a list of events that have happened since the most recent event time.
+  #
+  # Because Azure event timestamps are not necessarily stamped in order, an
+  # issue occurs where we could accidentally skip over events that happen
+  # in quick succession. We must therefore begin our query a couple minutes
+  # back from our most recent timestamp, and filter out any duplicates.
+  #
+  # See https://bugzilla.redhat.com/show_bug.cgi?id=1724312 for details.
+  #
   def get_events
-    # Grab only events for the last minute if this is the first poll
-    filter = @since ? "eventTimestamp ge #{@since}" : "eventTimestamp ge #{startup_interval}"
-    fields = 'authorization,description,eventName,eventTimestamp,resourceGroupName,resourceProviderName,resourceId,resourceType'
-    events = connection.list(:filter => filter, :select => fields, :all => true).sort_by(&:event_timestamp)
+    filter = "eventTimestamp ge #{most_recent_time}"
+
+    events = connection.list(:filter => filter, :select => SELECT_FIELDS, :all => true)
+
+    if events.present?
+      existing_records = EventStream.select(:ems_ref).where(:source => 'AZURE', :ems_ref => events.map(&:event_data_id)).map(&:ems_ref)
+      events = events.reject{ |e| existing_records.include?(e.event_data_id) } if existing_records.present?
```

Comment:
If you're just trying to prevent saving duplicate events we should be able to do that in the event handler.

There already is logic there to prevent duplicate events being created, https://github.com/ManageIQ/manageiq/blob/master/app/models/ems_event.rb#L188-L194

Reply:
@agrare Oh, how do we use that exactly? I see our Runner uses `EmsEvent.add_queue` at:

 https://github.com/ManageIQ/manageiq-providers-azure/blob/master/app/models/manageiq/providers/azure/cloud_manager/event_catcher/runner.rb#L29

Just change it to this?

`EmsEvent.add_queue('add', @cfg[:ems_id], EmsEvent.create(event_hash))`

Reply:
It _should_ happen automatically, since `EmsEvent.add_queue` calls `EmsEvent.add` which in turn calls `.create_event`

Reply:
Oops, that's a private method, hmm...

Reply:
@djberg96 were you seeing duplicate events get created or was this a "just in case" thing?

Reply:
As per our gitter conversation, this ultimately turned out to be an issue with timestamp precision and comparison.

Reply:
As per gitter, we were getting duplicates because of timestamp comparison failures. Fixed now.